### DISCUSSION TITLE
Split mapfile (.osu) into 2 parts

### DISCUSSION THREAD

**Discussion by Givikap120** [ppy/osu#27774] [Ideas]
.osu mapfile should be split into 2 parts: metadata and gameplay
with gameplay part potentially being binary file with already parsed and ready-to-use gameplay
there are several major benefits to this:
- you can edit map's metadata "on the fly" without reuploading map, disqualifying or maybe even unranking the map
- the replay is bound only to gameplay file hash, so you won't invalidate all replays by changing minor things
- binary gameplay file allows much faster beatmap loading and difficulty calculation, as right now map parsing and processing takes half of the difficulty calculation time 

about hacky maps where mapper often editing gameplay objects thru notepad:
you can export gameplay part into text file to edit and then import it again + generate binary file (binary file should be also generated for already existing maps)

**Comment by Walavouchey** [comment]
reworking the beatmap format is planned https://github.com/ppy/osu/discussions/12976#discussioncomment-791080, but it will take quite a while until it happens

**Reply by Givikap120** [reply]
considering that real-time difficulty processing is in pretty high priority rn, having preprocessed map binary is vital to achieve maximal speed
so for this reason from my perspective this feature is pretty important

**Reply by Rekunan** [reply]
> pretty high priority

The only place I see any mention of this is an [item](https://github.com/orgs/ppy/projects/19/views/1?pane=issue&itemId=57213728) in the pp release issue tracker, which is just a statement that they're going to explore it. It doesn't even have any priority label on it, so it's misleading to say that it's pretty high priority.

But regardless, having a map binary wouldn't be necessary as most of the speed is taken up by the actual calculations, not preprocessing.

**Reply by Givikap120** [reply]
you can profile diffcalc by yourself and confirm that >70% of the time is taken by different kinds of preprocessing

**Reply by Rekunan** [reply]
I actually don't know how to do that, can u tell me how?

**Reply by Givikap120** [reply]
the easiest way is to run osu-lazer in profiling mode (Visual Studio -> Debug -> Profiling)
then quickly go in the map list and select random mods to trigger star rating recalculation
do this with different maps for like a minute and then leave
with profiling report you can compare runtime for different functions of diffcalc
